### PR TITLE
Specification for tokens and create/mint/burn 

### PR DESCRIPTION
**Pull request by hansl**
This specification goes into details of what a token should contain
as information, update the ledger information regarding it (and
returns token informations instead of just names), and add a new
attribute for creating / minting / burning tokens.

It also includes account roles so a token owner can be an account
and contains roles for others.

### CODE REVIEW

File: attributes/network/cddl/2_ledger.cddl [hunk 1]
```diff
@@ -10,49 +10,119 @@ ledger-symbol = address
...
-            1 => tstr,  ; Description
-            2 => uint, ; Number of precision digits
-            tstr => tstr ; metadata
-        }
-    }
+    ; Binary content of the image.
+    2 => bstr,
+
+    ; Optional height and width, in pixels.
+    ? 3 => [ uint, uint ],
```
Comment (fmorency) [inline attributes/network/cddl/2_ledger.cddl:44]: Those can be extracted from the image header.

File: attributes/network/cddl/2_ledger.cddl [hunk 2]
```diff
@@ -10,49 +10,119 @@ ledger-symbol = address
...
+
+    ; Optional DPI.
+    ? 4 => uint,
```
Comment (fmorency) [inline attributes/network/cddl/2_ledger.cddl:47]: What will this be used for?

File: attributes/network/cddl/2_ledger.cddl [hunk 3]
```diff
@@ -10,49 +10,119 @@ ledger-symbol = address
...
+    ; mechanism, that would be less, as those tokens are not burnt but
+    ; frozen.
+    4 => uint,
+
+    ; The maximum supply for the token on this particular ledger.
+    ; This is the maximum amount of tokens at any point in time, including
+    ; in the far future, and should never change.
+    ; If this is null, then the token has an infinite maximum supply.
+    5 => uint / null,
+}
```
Comment (fmorency) [inline attributes/network/cddl/2_ledger.cddl:106]: Why are the indices starting at 3?

File: attributes/network/cddl/2_ledger.cddl [hunk 4]
```diff
@@ -10,49 +10,119 @@ ledger-symbol = address
...
+    ; be unique across all the MANYverse.
+    0 => ledger-symbol,
+
+    ; Summary information for the token.
+    1 => token-info-summary,
+
+    ; Supply information for the token.
+    2 => token-info-supply,
+
+    ; Owner of the token. If missing, the token has no owner.
```
Comment (fmorency) [inline attributes/network/cddl/2_ledger.cddl:120]: The token owner is mandatory in `tokens.create`, i.e., if the field is not supplied then the sender is used as the owner.

File: attributes/network/cddl/2_ledger.cddl [hunk 5]
```diff
@@ -10,49 +10,119 @@ ledger-symbol = address
...
+    ? 3 => non-anonymous-address,
+
+    ; A visual logo. If this is missing, a frontend should try to display
+    ; something in exchange.
```
Comment (fmorency) [inline attributes/network/cddl/2_ledger.cddl:124]:
```suggestion
    ; A visual logo. If this is missing, a frontend should display a placeholder
```

**Review by fmorency — CHANGES_REQUESTED** [review]
Thanks for this! Some comments/questions.

File: attributes/network/11_ledger_tokens.adoc [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+[sources,cddl]
+----
+include::{cddl}/11_ledger_tokens.cddl[tag=update]
+----
+
+== Minting and Burning
+
+The `tokens.mint` endpoint mints new tokens in the ledger associated with certain addresses instantly.
+Servers MUST return an error if minting the tokens means increasing the total supply over the maximum supply.
```
[the remark this answers is not in the document]
Comment (hansl) [reply attributes/network/11_ledger_tokens.adoc:32]: I'll keep the `Servers MUST return an error if minting the tokens means increasing the total supply over the maximum supply.` mention.

File: attributes/network/11_ledger_tokens.adoc [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+The `tokens.burn` endpoint burns new tokens in the ledger associated with certain addresses instantly.
+If a certain flag is passed, servers MUST return an error if accounts have fewer tokens than being burnt.
+Otherwise, servers MUST burn all tokens in the accounts and leave those accounts empty.
+
+Servers MAY delay minting and burning (for example, to process them in order at the end of the day).
+In this case, the command should be async and only resolved when the minting or burning actually occurs.
+See xref:../response/1_async.adoc[Async Responses] for more information on asynchronous operations.
```
[the remark this answers is not in the document]
Comment (hansl) [reply attributes/network/11_ledger_tokens.adoc]: Reworded. PTAL. `async` in this context means that the answer to the operation isn't available right away (you get an async token), hence the xref.

File: attributes/network/6_ledger_commands.adoc [hunk 8]
```diff
@@ -3,7 +3,25 @@
 
 This attribute is used by servers that supports commands related to ledger.
 
+== Events
+
+Calling `ledger.send` MUST add the `[6, 0]` event to the event log.
```
[the remark this answers is not in the document]
Comment (hansl) [reply attributes/network/6_ledger_commands.adoc]: Reworded.

File: attributes/network/cddl/11_ledger_tokens.cddl [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+; A token genesis event.
+ledger-tokens-event-info-create = {
+    0 => ledger-tokens-event-type-create,
+
+    1 => token-info-summary,            ; Summary.
+    2 => address,                       ; Symbol.
+    3 => address,                       ; Owner.
+    ? 4 => visual-token-logo,           ; Visual logo.
+    ? 5 => ledger-tokens-address-map,   ; Holders at genesis.
```
[the remark this answers is not in the document]
Comment (hansl) [reply attributes/network/cddl/11_ledger_tokens.cddl]: Yes, but I'll use "creation" for clarity.

File: attributes/network/cddl/11_ledger_tokens.cddl [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    ? 6 => ledger-amount,               ; Maximum supply, if any.
+}
+
+; An update to a token information.
+ledger-tokens-event-info-update = {
+    0 => ledger-tokens-event-type-update,
+    ? 0 => tstr,                        ; Token name.
+    ? 1 => tstr,                        ; Human readable ticker symbol.
```
[the remark this answers is not in the document]
Comment (hansl) [reply attributes/network/cddl/11_ledger_tokens.cddl]: https://learn.robinhood.com/articles/79dhnyd0hycqflR5un29us/what-is-a-ticker-symbol/

Basically short name for a token (like BTC for Bitcoin). I added more text to clarify.

File: attributes/network/cddl/11_ledger_tokens.cddl [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+    ? 2 => uint,                        ; Precision decimals.
+    ? 3 => transaction-fee,             ; Transaction fees.
+    ? 4 => memo,                        ; Memo of the token itself.
+    ? 5 => memo,                        ; Memo.
```
[the remark this answers is not in the document]
Comment (hansl) [reply attributes/network/cddl/11_ledger_tokens.cddl]: We do (it's in token-info-summary#4). This is to update the one we have at creation.

File: attributes/network/cddl/11_ledger_tokens.cddl [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+tokens.create@param = {
+    ; The summary information for the token.
+    0 => token-info-summary,
+
+    ; An optional owner for the token. If missing, the message sender will be
+    ; used as owner. There can only be one owner.
+    ? 1 => non-anonymous-address,
+
+    ; An optional visual logo. See 2_ledger.cddl for more information on logos.
+    ? 2 => visual-token-logo,
```
[the remark this answers is not in the document]
Comment (hansl) [reply attributes/network/cddl/11_ledger_tokens.cddl:71]: Probably. Let me think about this.

File: attributes/network/cddl/11_ledger_tokens.cddl [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    ; Servers MAY error if this field is updated or out of supported range.
+    ? 2 => uint,
+
+    ; Transaction fees.
+    ; Servers MAY error if this field is updated to invalid values.
+    ? 3 => transaction-fee,
+
+    ; A memo associated with the token. This will override the memo, not just
+    ; add to it.
+    ? 4 => memo,
```
[the remark this answers is not in the document]
Comment (hansl) [reply attributes/network/cddl/11_ledger_tokens.cddl:111]: Sort of human and machine readable description. Like a README, references to URLs and company names, that kind of stuff.

File: attributes/network/cddl/2_ledger.cddl (hunk 1 above)
Comment (hansl) [reply attributes/network/cddl/2_ledger.cddl]: It can actually be different in some cases (pixels vs points). https://blog.fluidui.com/designing-for-mobile-101-pixels-points-and-resolutions/

Maybe I shouldn't dig too much into that.

File: attributes/network/cddl/2_ledger.cddl (hunk 2 above)
Comment (hansl) [reply attributes/network/cddl/2_ledger.cddl]: I was going to give some metadata that could be useful without reading the image, but I guess that's too much. Removed.

File: attributes/network/cddl/2_ledger.cddl (hunk 3 above)
Comment (hansl) [reply attributes/network/cddl/2_ledger.cddl]: At first it was merged with summary, but not anymore so removed.

File: attributes/network/cddl/2_ledger.cddl (hunk 4 above)
Comment (hansl) [reply attributes/network/cddl/2_ledger.cddl]: Yeah I made it optional.

File: attributes/network/cddl/2_ledger.cddl (hunk 1 above)
Comment (fmorency) [reply attributes/network/cddl/2_ledger.cddl]: I wouldn't. Let the front end deal with the data.

File: attributes/network/11_ledger_tokens.adoc [hunk 14]
```diff
@@ -30,13 +30,27 @@ include::{cddl}/11_ledger_tokens.cddl[tag=update]
 
 One can update a token to not have an owner by giving ownership to `null` (NOT anonymous).
 
+=== Extended Information
+
+A token can have optional and separately defined extended information.
+This information is specified in separate files and can be extended in the future.
+
+Events in the event log do not contain extended information, as these can be large in size.
```
Comment (fmorency) [inline attributes/network/11_ledger_tokens.adoc:38]: I would rephrase. 

The extended information attribute indices are present in the event log, not the associated data.

File: attributes/network/11_tokens/cddl/1_visual_logo.cddl [hunk 15]
```diff
@@ -0,0 +1,45 @@
+; A single logo that can be shown for representing the token. A token can
```
Comment (fmorency) [inline attributes/network/11_tokens/cddl/1_visual_logo.cddl:1]: The parent folder should probably be named `11_ledger_tokens` to match the attribute name.

File: attributes/network/cddl/11_ledger_tokens.cddl [hunk 16]
```diff
@@ -106,19 +143,18 @@ tokens.update@param = {
...
 
     ; A memo associated with this update.
-    ? 7 => memo,
+    ? 5 => memo,
+
+    ; Extended info to add. This is merged with the existing extended info.
+    ? 6 => token-extended-info,
+
+    ; Extended info to remove.
+    ? 7 => [ + attribute-related-index ],
```
Comment (fmorency) [inline attributes/network/cddl/11_ledger_tokens.cddl:157]: Out of curiosity, why did you decide to use two fields?

**Review by fmorency — CHANGES_REQUESTED** [review]
Minor changes and a question. The rest looks good, thanks!

File: attributes/network/11_tokens/cddl/1_visual_logo.cddl (hunk 15 above)
Comment (hansl) [reply attributes/network/11_tokens/cddl/1_visual_logo.cddl:1]: Done, but we haven't been very strict about it so far. Would probably be useful to refactor the others in a separate PR.

File: attributes/network/cddl/11_ledger_tokens.cddl (hunk 16 above)
Comment (hansl) [reply attributes/network/cddl/11_ledger_tokens.cddl:157]: You might want to add new info, or delete some infos, without touching the rest (since it could be a lot of data). How would you see this otherwise? A record containing two fields?

Comment (fmorency) [reply attributes/network/cddl/11_ledger_tokens.cddl:157]: As discussed, I would try to be consistent with the rest of the spec, either by re-passing the whole list, like we do for every other field we want to update, or by adding new endpoints as we do with features.

**Review by fmorency — APPROVED** [review]
LGTM! Thanks!

File: attributes/network/11_ledger_tokens/cddl/1_visual_logo.cddl [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+single-visual-token-logo = single-visual-token-logo-unicode
+                         / single-visual-token-logo-image
+
+; A unicode logo.
+; For Bitcoin this would be U+20BF ("₿").
+single-visual-token-logo-unicode = {
+    ; Type.
+    0 => 0,
+
+    ; Unicode code point. This isn't limited to a single glyph but can be
```
Comment (fmorency) [inline attributes/network/11_ledger_tokens/cddl/1_visual_logo.cddl:12]: I would rephrase to make it clear that we don't actually want the Unicode code point here.

File: attributes/network/cddl/12_ledger_mintburn.cddl [hunk 18]
```diff
@@ -0,0 +1,70 @@
+; tag::events[]
+event-info /= ledger-tokens-event-info-create
+            / ledger-tokens-event-info-mint
+            / ledger-tokens-event-info-burn
+
+ledger-tokens-event-type-mint = [11, 4]
+ledger-tokens-event-type-burn = [11, 5]
```
Comment (fmorency) [inline attributes/network/cddl/12_ledger_mintburn.cddl:7]:
```suggestion
ledger-tokens-event-type-mint = [12, 0]
ledger-tokens-event-type-burn = [12, 1]
```

File: attributes/network/11_ledger_tokens.adoc [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+:cddl: ./cddl/
+
+This attribute specifies that a ledger can have tokens managed by an authority.
+Without this attribute, tokens are defined at the genesis of the ledger, and cannot be created, minted, or burnt.
+
+Tokens described by this attribute are fungible and well specified.
+They would be equivalent to ERC-20 tokens.
+
+A server implementing this attribute SHOULD protect the endpoints described in this form in some way.
+For example, endpoints SHOULD error if the sender isn't from a certain address.
```
Comment (fmorency) [inline attributes/network/11_ledger_tokens.adoc:11]: How should the new symbol (Address) be generated?

File: attributes/network/11_ledger_tokens.adoc [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+== Creating Tokens
+
+One can create new tokens using the `tokens.create` command.
+
+[sources,cddl]
+----
+include::{cddl}/11_ledger_tokens.cddl[tag=create]
+----
+
+One can update the token information using the `tokens.update` command.
```
Comment (fmorency) [inline attributes/network/11_ledger_tokens.adoc:22]: Why can update a token? What is the permission scheme?

File: attributes/network/9_account_features/cddl/3_account_tokens.cddl [hunk 21]
```diff
@@ -0,0 +1,5 @@
+; tag::types[]
+account-role /= "canTokensMint"
+             /  "canTokensBurn"
+             /  "canTokensUpdate"
```
Comment (fmorency) [inline attributes/network/9_account_features/cddl/3_account_tokens.cddl:4]:
```suggestion
account-role /= "canTokensMint"
             /  "canTokensBurn"
             /  "canTokensUpdate"
             /  "canAddExtendedInfo"
             /  "canRemoveExtendedInfo"
```

File: attributes/network/11_ledger_tokens.adoc (hunk 19 above)
Comment (hansl) [reply attributes/network/11_ledger_tokens.adoc:11]: Clarified.

File: attributes/network/11_ledger_tokens.adoc (hunk 20 above)
Comment (hansl) [reply attributes/network/11_ledger_tokens.adoc:27]: Up to the server.

**Review by hansl — COMMENTED** [review]
PTAL.
